challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,294 @@
+  work we're doing here.
+
+We believe this technique is not only better, but also is more fun compared to
+whiteboard/quiz interviews so common in the industry. It's not without the
+downsides - it could take longer than traditional interviews.
+
+[Some of the best teams use coding challenges.](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/)
+
+We appreciate your time and are looking forward to hack on this project
+together.
+
+# Levels
+
+There are 6 engineering levels at Teleport. It's possible to score on level 1-5
+through coding challenge.
+
+Level 6 is only for internal promotions. Check the [engineering levels
+document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
+for more details.
+
+# Interview process
+
+The interview process will start with you receiving an invite to a private
+Slack channel. That channel will contain the interview panel. You can ask them
+about the engineering culture, work and life balance, or anything else that you
+would like to learn about Teleport.
+
+Before writing any actual code, we ask that you write a brief design document.
+The design document should cover: design approach, scope, proposed API,
+security considerations, CLI UX, and implementation details where appropriate.
+
+Please submit the design document as Markdown and PR#1 to your repository to
+allow us to provide you feedback on the proposed design.
+
+A few notes about the design document:
+
+* We expect the design document to be complete roughly within the first week.
+  This is to ensure you have enough time to work on the implementation.
+* Avoid writing an overly detailed design document. Two to three pages is
+  sufficient.
+* Avoid sending us draft design documents, it is difficult to evaluate which
+  parts are draft and which parts are complete. Instead we encourage asking
+  questions in Slack and sharing a design document that is ready to be reviewed.
+
+Once the design document has been approved by two reviewers, move onto
+implementation.
+
+Split your code submission into roughly 3-5 Pull Requests to give the team an
+opportunity to review your code and provide feedback. Feel free to merge each
+PR after you have two approvals.
+
+Our team will do their best to provide a high quality review of the submitted
+Pull Requests in a reasonable time frame. You are spending your time on this,
+we are going to contribute our time too.
+
+After the final submission, the interview team will assemble and vote using a
+"+1, -2" anonymous voting system: +1 is submitted whenever a team member
+accepts the submission, -2 otherwise.
+
+In case of a positive result, we will connect you to our HR and recruiting
+teams, who will work out the details and present an offer.
+
+In case of a negative score result, hiring manager will contact you and share a
+list of the key observations from the team that affected the result.
+
+# Requirements
+
+## Level 1
+
+### Library
```

Comment:
This is the first time we mention library. Do we actually care that candidates create a library that is separate from the CLI, or should we just let them organize their code as they see fit?

Reply:
You're right it doesn't need to be a library, but it felt like we needed some logical grouping for both the reviewers (for the early exit point) and for readability of the challenge (otherwise L5 becomes a giant list of bullet points).

What do you think?

Reply:
Let me think about this a bit more.

If we want a library, we should add more info up front rather than casually mentioning it in the title. But maybe I can think of a better way to have an early exit point.